### DISCUSSION TITLE
Is there a os.utime() that doesn't follow links?

### DISCUSSION THREAD

**Discussion by mike-ward** [vlang/v#21508] [Questions and Answers] [answered]
I'm writing a `touch` program for the CoreUtils project. It appears `os.utime()` follows links to their source. Is there a vlib api that can set the time of symbolic links? I need it to implement the `-h` option of `touch`

**Comment by mike-ward** [comment]
I took a shot at writing a `lutime` function. It changes the access and mod times of the given file when I run it, but the times are wildly wrong. I'm not a C programmer so I'm stumped why this doesn't work (Mac Sonoma 14.4)

```
#include <sys/time.h>

fn C.lutimes(&char, voidptr) int

fn lutime(path string, actime int, modtime int) ! {
	mut u := [C.timeval{u64(actime), 0}, C.timeval{u64(modtime), 0}]
	result := C.lutimes(&char(path.str), voidptr(&u))
	if result != 0 {
		return error('lutime failed (${C.errno})')
	}
}
```

https://nxmnpg.lemoda.net/2/utimes

**Comment by mike-ward** [comment, marked as answer]
I was able to resolve the issue by using the `data` member of `Array`.

```
fn lutime(path string, actime int, modtime int) ! {
 	times := [C.timeval{u64(actime), u64(0)}, C.timeval{u64(modtime), u64(0)}]
	if C.lutimes(&char(path.str), voidptr(times.data)) != 0 {
		return error('lutime failed (${C.errno})')
	}
}
```

**Comment by JalonSolov** [comment]
The alternative would be to use a fixed array, which is only data.  Since this array size is known, and will never change in size, fixed is likely better here.
```v
 	times := [C.timeval{u64(actime), u64(0)}, C.timeval{u64(modtime), u64(0)}]!
 	if C.lutimes(&char(path.str), voidptr(&times)) != 0 {
		return error('lutime failed (${C.errno})')
	}
```

**Reply by mike-ward** [reply]
Fixed array generates a warning that suggests using `&times[0]`. (Excellent error message and diagnostic by the way)

```
touch.v:146:32: warning: cannot cast a fixed array (use e.g. `&arr[0]` instead)
  144 | fn lutime(path string, acctime int, modtime int) ! {
  145 |     times := [C.timeval{u64(acctime), u64(0)}, C.timeval{u64(modtime), u64(0)}]!
  146 |     if C.lutimes(&char(path.str), voidptr(times)) != 0 {
      |                                   ~~~~~~~~~~~~~~
```

**Reply by JalonSolov** [reply]
Ah, right.  Weird the way that works, but... it does as long as you do what V wants.  ;-)
